### PR TITLE
SILGen: Short-cut bind_optional_exprs that are used by objective-c pr…

### PR DESCRIPTION
**Pull request by aschwaighofer**
…operty and method calls

obj?.method() // We may just forward the value of obj to method()

Objective-C's message send returns nil for a message send to nil so we
can skip the enum switch on the optional value and use it as is.

rdar://48007302

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

### CODE REVIEW

**Comment by aschwaighofer** [comment]
@swift-ci Please test

File: lib/SILGen/SILGenExpr.cpp [hunk 1]
```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
...
+  return nullptr;
+}
+
 RValue RValueEmitter::visitMemberRefExpr(MemberRefExpr *E, SGFContext C) {
   assert(!E->getType()->is<LValueType>() &&
          "RValueEmitter shouldn't be called on lvalues");
 
+
   if (isa<TypeDecl>(E->getMember().getDecl())) {
+    MarkBindOptionalShortCut shortCut(SGF, E);
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: This is for a member access that is a type, eg something like type(of: value).SomeNestedClass. This doesn’t apply here.

File: lib/SILGen/SILGenExpr.cpp [hunk 2]
```diff
@@ -4438,10 +4540,14 @@ void SILGenFunction::emitBindOptionalAddress(SILLocation loc,
   // Since we know that we have an address, we do not need to worry about
   // ownership invariants. Instead just use a select_enum_addr.
   SILBasicBlock *someBB = createBasicBlock();
+  if (mayShortCut) {
+    B.createBranch(loc, someBB);
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: You don’t need the basic block at all in this case I think. It should be enough to just bitcast the optional to its payload type.

File: lib/SILGen/SILGenLValue.cpp [hunk 3]
```diff
@@ -3280,12 +3282,25 @@ LValue SILGenLValue::visitBindOptionalExpr(BindOptionalExpr *e,
...
   // Project out the payload on the success branch.  We can just use a
   // naked ValueComponent here; this is effectively a separate l-value.
-  ManagedValue valueAddr =
-    getAddressOfOptionalValue(SGF, e, optAddr, valueTypeData);
+  ManagedValue valueAddr;
+  if (mayShortCut) {
+    bool hadCleanup = optAddr.hasCleanup();
+    SILValue addr = optAddr.forward(SGF);
+    auto eltTy = addr->getType()
+                     .getObjectType()
```
Comment (slavapestov) [inline lib/SILGen/SILGenLValue.cpp]: I don’t think the getObjectType() is necessary

File: lib/SILGen/SILGenExpr.cpp [hunk 4]
```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
...
 } // end anonymous namespace
 
+// We may simply forward the Optional<SomeClass> value as SomeClass if the
+// receiver of the value is the self argument of objc_msgSend.
+BindOptionalExpr *SILGenFunction::maySkipSwitchInBindOptionalSubExpr(Expr *E) {
+
+  // Acceptable output types are class types and void (since it does not have
+  // uses), and optional thereof. Those types can be satisfied by the return
+  // value of the objc_msgSend without change in representation.
+  auto hasAcceptableOuputType = [](Expr *expr) -> bool {
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: Typo: Ouput not Output

File: lib/SILGen/SILGenExpr.cpp [hunk 5]
```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
...
+    if (auto *load = dyn_cast<LoadExpr>(self))
+      self = load->getSubExpr();
+    self = dyn_cast<BindOptionalExpr>(self);
+    if (!self)
+      return nullptr;
+
+    auto *func = dyn_cast<DeclRefExpr>(dotCall->getFn());
+    if (!func)
+      return nullptr;
+    if (!func->getDecl()->isObjC())
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: isObjC() is not the same as “called via objc_msgSend”

File: lib/SILGen/SILGenExpr.cpp [hunk 6]
```diff
@@ -2149,11 +2150,109 @@ class NominalTypeMemberRefRValueEmitter {
 
 } // end anonymous namespace
 
+// We may simply forward the Optional<SomeClass> value as SomeClass if the
+// receiver of the value is the self argument of objc_msgSend.
+BindOptionalExpr *SILGenFunction::maySkipSwitchInBindOptionalSubExpr(Expr *E) {
```
Comment (slavapestov) [inline lib/SILGen/SILGenExpr.cpp]: All the checks here are duplicating existing logic elsewhere in SILGen that tries to determine if a method is dispatched via objc_msgSend(). This is going to be a source of latent bugs.

**Review by slavapestov — CHANGES_REQUESTED** [review]
I really don’t like this change. Please redo it as a SIL optimizer pass as John points out below.

**Comment by slavapestov** [comment]
We should instead do this at the ArgumentSource level when lowering a call to a method or accessor. We already know at that point if it’s an Objective C message send.

**Comment by rjmccall** [comment]
This is not generally semantically valid; ?-chains that short-circuit are required to bypass the evaluation of all expression to the right of the `?` operator within the chain, including call arguments.

I would recommend applying this optimization at the SIL level, where it will also avoid complicating an already over-complex bit of code in SILGen.

**Comment by rjmccall** [comment]
Doing it at the SIL level would also let you optimize other syntactic forms of short-circuiting, like `guard let`.

**Pull request closed without merging**